### PR TITLE
Provide a way to set a route instead of the absolute link in a select or action column

### PR DESCRIPTION
**Pull request by halfpastfouram**
Solution for issue #268.

### Solution for Action Button
I've used the ZF2 view helper URL inside `ZfcDatagrid\Renderer\BootstrapTable\View\Helper\TableRow` since that is already a view helper itself. If you've got a router setup this works out of the box.

You can now do the following:

```php
$action->setRoute( 'myRoute' );
$action->setRouteParams( [
    'month' => $column->getColumnValuePlaceholder( 'month' ),
    'id'    => $column->getColumnValuePlaceholder( 'id' ),
] );
```

### Solution for Select Column
Set the router you want to use in the Datagrid:

```php
$datagrid->setRouter( $serviceLocator->get( 'router' ) );
```

Then set a route on a link formatter and provide that formatter to the column:

```php
$formatter = new \ZfcDatagrid\Column\Formatter\Link();
$formatter->setRoute( 'myroute' );
$formatter->setRouteParams( [
    'month' => $column->getColumnValuePlaceholder( 'month' ),
    'id'    => $column->getColumnValuePlaceholder( 'id' ),
] );
$col1->addFormatter( $formatter );
```

### CODE REVIEW

File: src/ZfcDatagrid/Datagrid.php [hunk 1]
```diff
@@ -399,6 +405,30 @@ public function hasTranslator()
...
+    {
+        return $this->router;
+    }
+
+    /**
+     * @return bool
+     */
+    public function hasRouter()
+    {
+        return !is_null($this->translator);
```
Comment (ThaDafinser) [inline src/ZfcDatagrid/Datagrid.php]: @halfpastfouram wrong variable

File: src/ZfcDatagrid/PrepareData.php [hunk 2]
```diff
@@ -225,6 +247,12 @@ public function prepare()
                  */
                 if ($col->hasFormatters() === true) {
                     foreach ($col->getFormatters() as $formatter) {
+                        if ($formatter instanceof Column\Formatter\RouterInterface
+                            && $this->getRouter() instanceof RouteStackInterface
+                        ) {
+                            /** @var Column\Formatter\RouterInterface */
+                            $formatter->setRouter($this->getrouter());
```
Comment (ThaDafinser) [inline src/ZfcDatagrid/PrepareData.php]: @halfpastfouram method lower case

**Comment by ThaDafinser** [comment]
@halfpastfouram the `prefer-lowest` parts i failing because of the resolving

In the other there are some `php-cs-fixer` issues
just run `vendor/bin/php-cs-fixer fix` and they should turn green

https://travis-ci.org/ThaDafinser/ZfcDatagrid/jobs/201881803#L328

**Comment by ThaDafinser** [comment]
@halfpastfouram after the cs issues are resolved, i'm going to fix the resolve issue.

In general i would say the PR LGTM - a small improvement would be a `trait` / `interface` for the get/set/hasRouter part, since it's "copied" several times - but it's not required.

**Comment by halfpastfouram** [comment]
@ThaDafinser I agree with you on the trait and interface. Would indeed be an improvement, but haven't had the time to implement that yet. I could improve it a bit in a later stage. Right now I can't wait until this is implemented because it's gonna make my life/job so much easier.

What do you mean exactly by the cs issues? Is there anything you need me to do?

**Comment by ThaDafinser** [comment]
@halfpastfouram i will fix it for you in your PR, since you don't know it (later this day)
https://github.com/FriendsOfPHP/PHP-CS-Fixer

It's just coding style violations

**Comment by ThaDafinser** [comment]
merged to `develop` branch see https://github.com/ThaDafinser/ZfcDatagrid/pull/272

Thank you! :+1: 
(your contributions will stay intact over your commits)

**Pull request closed without merging**